frame/support/test/tests/pallet_version.rs

```diff
@@ -0,0 +1,170 @@
+// This file is part of Substrate.
+
+// Copyright (C) 2020 Parity Technologies (UK) Ltd.
+// SPDX-License-Identifier: Apache-2.0
+
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+// 	http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+//! Tests related to the pallet version.
+
+#![recursion_limit="128"]
+
+use codec::{Decode, Encode};
+use sp_runtime::{generic, traits::{BlakeTwo256, Block as _, Verify}};
+use frame_support::{
+	traits::{PALLET_VERSION_STORAGE_KEY_POSTFIX, PalletVersion, OnRuntimeUpgrade},
+	crate_to_pallet_version, weights::Weight,
+};
+use sp_core::{H256, sr25519};
+
+mod system;
+
+/// A version that we will check for in the tests
+const SOME_TEST_VERSION: PalletVersion = PalletVersion { major: 3000, minor: 30, patch: 13 };
+
+/// Checks that `on_runtime_upgrade` sets the latest pallet version when being called without
+/// being provided by the user.
+mod module1 {
+	use super::*;
+
+	pub trait Trait: system::Trait {}
+
+	frame_support::decl_module! {
+		pub struct Module<T: Trait> for enum Call where
+			origin: <T as system::Trait>::Origin,
+			system = system,
+		{}
+	}
+}
+
+/// Checks that `on_runtime_upgrade` sets the latest pallet version when being called and also
+/// being provided by the user.
+mod module2 {
+	use super::*;
+
+	pub trait Trait<I=DefaultInstance>: system::Trait {}
+
+	frame_support::decl_module! {
+		pub struct Module<T: Trait<I>, I: Instance=DefaultInstance> for enum Call where
+			origin: <T as system::Trait>::Origin,
+			system = system
+		{
+			fn on_runtime_upgrade() -> Weight {
+				assert_eq!(crate_to_pallet_version!(), Self::current_version());
+
+				let version_key = PalletVersion::storage_key::<T::PalletInfo, Self>().unwrap();
+				let version_value = sp_io::storage::get(&version_key);
+
+				if version_value.is_some() {
+					assert_eq!(SOME_TEST_VERSION, Self::storage_version());
+				} else {
+					// As the storage version does not exist yet, it should be the same as
+					// the current version.
+					assert_eq!(Self::storage_version(), Self::current_version());
+				}
+
+				0
+			}
+		}
+	}
+
+	frame_support::decl_storage! {
+		trait Store for Module<T: Trait<I>, I: Instance=DefaultInstance> as Module2 {}
+	}
+}
+
+impl module1::Trait for Runtime {}
+impl module2::Trait for Runtime {}
+impl module2::Trait<module2::Instance1> for Runtime {}
+impl module2::Trait<module2::Instance2> for Runtime {}
```

Comment:
Not related to the PR but I'm realizing someones can implement the trait for Runtime and use it, but forget to put in construct_runtime.
This will lead to various errors where PalletInfo will fail to return name. But I'm not sure how to improve this.

Reply:
But when someone only implements the trait, nothing of the pallet will probably be called? 

In the end a lot of stuff will fail when you don't put it into `construct_runtime`